File: backend/app/schemas/document.py
"""Document API schemas."""

from datetime import datetime

from pydantic import BaseModel, Field


# ── Upload / response shapes ──────────────────────────────────────────────────

class DocumentUploadResponse(BaseModel):
    """Returned immediately after a successful file upload."""
    document_id: str
    filename: str
    file_type: str
    file_size_bytes: int
    status: str                      # "uploaded"
    job_id: str                      # Celery task ID for polling


class DocumentStatusResponse(BaseModel):
    """Polling response for a single document's processing state."""
    document_id: str
    filename: str
    status: str                      # uploaded | processing | extracting | indexing | completed | error
    progress: int = Field(ge=0, le=100)   # 0-100
    error_message: str | None = None
    chunk_count: int = 0
    created_at: datetime
    updated_at: datetime


class DocumentSummarySchema(BaseModel):
    """Brief document entry for list views."""
    id: str
    filename: str
    file_type: str
    file_size_bytes: int
    status: str
    chunk_count: int = 0
    supplier_id: str | None = None
    created_at: datetime


class DocumentListSchema(BaseModel):
    """Paginated list of documents."""
    items: list[DocumentSummarySchema] = Field(default_factory=list)
    total: int = 0


# ── Chunk schema ──────────────────────────────────────────────────────────────

class DocumentChunkSchema(BaseModel):
    """A single extracted text chunk, as stored in PostgreSQL."""
    id: str
    document_id: str
    chunk_index: int
    page_number: int
    section_name: str | None = None
    content: str
    token_count: int
    extraction_confidence: float

class DocumentChunkWithMetaSchema(DocumentChunkSchema):
    """Chunk with metadata (like document filename)."""
    document_filename: str
    created_at: datetime


# ── Job status (Celery task) ──────────────────────────────────────────────────

class JobStatusResponse(BaseModel):
    """
    Lightweight job-tracking response.

    Maps to the Celery task result stored in Redis.
    Mirrors the Document.status field so the frontend can use either
    endpoint for polling.
    """
    job_id: str
    document_id: str
    status: str          # pending | processing | extracting | indexing | completed | failed
    progress: int = Field(ge=0, le=100)
    detail: str = ""     # human-readable current step description
    error: str | None = None

File: frontend/src/types/document.ts
export interface DocumentUploadResponse {
  document_id: string;
  filename: string;
  file_type: string;
  file_size_bytes: number;
  status: string;
  job_id: string;
}

export type DocumentStatus =
  | 'uploaded'
  | 'processing'
  | 'extracting'
  | 'indexing'
  | 'completed'
  | 'error';

export interface DocumentStatusResponse {
  document_id: string;
  filename: string;
  status: DocumentStatus;
  progress: number;
  error_message: string | null;
  chunk_count: number;
  created_at: string;
  updated_at: string;
}

export interface DocumentSummary {
  id: string;
  filename: string;
  file_type: string;
  file_size_bytes: number;
  status: DocumentStatus;
  chunk_count: number;
  supplier_id: string | null;
  created_at: string;
}

export interface DocumentChunk {
  id: string;
  document_id: string;
  chunk_index: number;
  page_number: number;
  section_name: string | null;
  content: string;
  token_count: number;
  extraction_confidence: number;
}

export interface DocumentChunkWithMeta extends DocumentChunk {
  document_filename: string;
  created_at: string;
}

export interface JobStatusResponse {
  job_id: string;
  document_id: string;
  status: DocumentStatus | 'pending' | 'failed';
  progress: number;
  detail: string;
  error: string | null;
}
